packages/tpetra/core/src/Tpetra_LinearProblem_def.hpp

```diff
@@ -0,0 +1,192 @@
+    : dist_object_type (Problem),
+      Operator_(Problem.Operator_),
+      A_(Problem.A_),
+      X_(Problem.X_),
+      B_(Problem.B_)
+  {
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  leftScale(const Teuchos::RCP<const vector_type> & D, Teuchos::ETransp mode)
+  {
+    const Scalar ST0 = Teuchos::ScalarTraits<Scalar>::zero();
+    const Scalar ST1 = Teuchos::ScalarTraits<Scalar>::one();
+    if (mode == Teuchos::NO_TRANS) {
+      A_->leftScale(*D);
+      B_->elementWiseMultiply(ST1, *D, *B_, ST0);
+    }
+    else {
+      A_->rightScale(*D);
+      vector_type R(*D, Teuchos::DataAccess::Copy);
+      R.reciprocal(*D);
+      X_->elementWiseMultiply(ST1, R, *X_, ST0);
+    }
+  
+    return;
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  rightScale(const Teuchos::RCP<const vector_type> & D, Teuchos::ETransp mode)
+  {
+    const Scalar ST0 = Teuchos::ScalarTraits<Scalar>::zero();
+    const Scalar ST1 = Teuchos::ScalarTraits<Scalar>::one();
+    if (mode == Teuchos::NO_TRANS) {
+      A_->rightScale(*D);
+      vector_type R(*D, Teuchos::DataAccess::Copy);
+      R.reciprocal(*D);
+      X_->elementWiseMultiply(ST1, R, *X_, ST0);
+    }
+    else {
+      A_->leftScale(*D);
+      B_->elementWiseMultiply(ST1, *D, *B_, ST0);
+    }
+    return;
+  }
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  int LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  checkInput(bool fail_on_error) const {
+
+    int error = 0;
+    if (fail_on_error) {
+
+      const char tfecfFuncName[] = "checkInput: ";
+
+      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(Operator_==Teuchos::null, 
+        std::logic_error, "Operator_ is unset.");
+
+      TPETRA_ABUSE_WARNING(A_==Teuchos::null, std::runtime_error, 
+        "Linear problem does not have a matrix (A_), just an operator.");
+  
+      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(X_==Teuchos::null, 
+        std::logic_error, "Solution vector (X_) is unset.");
+
+      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(B_==Teuchos::null, 
+        std::logic_error, "RHS vector (B_) is unset.");
+  
+      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(!A_->getRowMap()->isSameAs(*(X_->getMap())),
+        std::logic_error, "Domain map of matrix is not the 'same as' the solution map.");
+
+      TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(!A_->getRowMap()->isSameAs(*(B_->getMap())),
+        std::logic_error, "Range map of matrix is not the 'same as' the RHS map.");
+    }
+    else {
+      if (Operator_==Teuchos::null) error = -1;
+      if (A_==Teuchos::null) error = 1; // Return warning error because this problem has no matrix (just an operator)
+      if (X_==Teuchos::null) error = -2;
+      if (B_==Teuchos::null) error = -3;
```

Comment:
@ccober6 I'm not a fan of returning error codes.  What about just throwing an exception or a hard failure?

Reply:
Yeah, same here and I initially started that way, but did not know what Tpetra's "policy" was, and what the stakeholder usage looked like to maintain similar functionality.  That's why I optionally added it.  I am open to removing the error codes.

Reply:
There's no backwards compatibility to worry about because this is new functionality.  I'll let others chime in, but I think throwing an exception is probably the way to go.

Reply:
Yeah, not true backward compatibility but work the same as Epetra_LinearProblem::checkInput().  Throwing is probably the best route.

Reply:
@jhux2, ok removed the error code.  Can I get another review?